sourced/ml/mining/cmd/extract_stdlib.py

```diff
@@ -0,0 +1,121 @@
+import argparse
+from io import BytesIO
+import json
+import logging
+from pathlib import Path
+import re
+from zipfile import ZipFile
+import zlib
+
+from lxml import html
+import requests
+from requests.exceptions import HTTPError
+import yaml
+
+from sourced.ml.mining.models import StandardLibraries
+from sourced.ml.mining.utils import check_remove_filepath, path_with_suffix
+
+
+def get_page(url: str):
+    """Return the GET response to the given URL or raise an error."""
+    try:
+        return requests.get(url)
+    except HTTPError as http_err:
+        print("HTTP error occurred: %s" % http_err)
+    except Exception as err:
+        print("Other error occurred: %s " % err)
+
+
+def get_leaves(url: str, xpath: str):
+    """Return the nodes extracted from the GET response's content."""
+    content = get_page(url).content
+    tree = html.fromstring(content)
+    return tree.xpath(xpath)
+
+
+def is_zlib_header(header: bytes):
+    """Decode the input header, return True if successful and False if not."""
+    try:
+        header.decode(encoding="utf-8")
+        return True
+    except UnicodeDecodeError:
+        return False
+
+
+def extract_stdlib(args: argparse.Namespace):
```

Comment:
Yes, indeed. Scrapy will make it muuuuch easier and more reliable, with errors handling, etc. Example: https://github.com/src-d/identity-matching/blob/master/research/names.py

Reply:
Scrapy is only for the crawling tasks ofc. Forget about xpath, nobody uses it IRL. It has CSS selectors and the same tree traversal capabilities as lxml. So instead of navigating a tree, you can `.my-css-class` and done.

The HTTP  stuff is very hard to do properly and reliably from scratch, so we should save tons of maintenance in the future.